text/0000-add-scratch-stack.md

```diff
@@ -0,0 +1,161 @@
+This is most common in golang, but is also often utilized in rust. It is also
+common in languages/toolchains not supported by Paketo like C/C++. These
+statically-linked binaries can be run on a minimally-modified empty container
+image (AKA "scratch" image) with few-to-no packages/libraries. Scratch images
+are a common choice of container image for statically-linked executables as it
+minimizes the size (in MB) of the image as well as the number of installed
+packages - in turn reducing the surface area for security issues.
+
+Currently, in order to build these statically-linked applications using Paketo
+buildpacks (and aiming for the smallest possible resultant run image) users
+must either opt for the [tiny
+stack](https://github.com/paketo-buildpacks/jammy-tiny-stack) or build their
+own scratch stack. We have evidence of users doing both of these things.
+Neither are satisfactory to users, however, as the tiny stack contains more
+packages than is necessary for those applications, and users would prefer not
+to maintain their own stacks if Paketo is willing and able to do so.
+
+Perhaps most importantly, we are [considering adding more packages to the tiny
+stack](https://github.com/paketo-buildpacks/rfcs/pull/231/files), and the
+biggest objection to that proposal is that it will further increase the size
+and surface area of the tiny image for statically-linked binaries. This could,
+in-turn, make Paketo less attractive for developers of these applications.
+
+Creating a scratch stack will decouple these use cases, allowing us to tailor
+the tiny stack to applications which want more packages, while also providing a
+better experience for statically-linked binaries.
+
+## Detailed Explanation
+
+### Repository/Image Location
+
+We will create a new Github repository at
+`paketo-buildpacks/jammy-scratch-stack` and a new pair of images on Dockerhub
+at `paketobuildpacks/build-jammy-scratch` and
+`paketobuildpacks/run-jammy-scratch`.
+
+We will adopt the tag system used for Jammy - specifically semver (i.e.
+'X.Y.Z') and 'latest' tags.
+
+We will not create a builder repository or image, though we reserve the option
+for the Builder team to do so in the future if that is requested and approved.
+
+### Stack ID
+
+We propose creating a new stack ID: `io.buildpacks.stacks.jammy.scratch`.
+This may require approval from the upstream Cloud-Native Buildpacks project.
+
+### Image Contents
+
+The scratch stack run image will contain the `tzdata` and `ca-certificates`
+packages, the standard CNB/Paketo metadata (including users and /etc/passwd
+entries), and nothing else. These two packages are often provided in scratch
+images as they are small and have minimal security risk.
+
+The build image will contain the same packages as the tiny stack (probably with
+the addition of the packages proposed in [the tiny packages
+RFCS](https://github.com/paketo-buildpacks/rfcs/pull/231/files)).
```

Comment:
Is there a clever way we could just reuse the tiny build image for the scratch build image, rather than making a duplicate?

Reply:
Ah, I guess this isn't that useful because we'll still need metadata changes.

Reply:
Currently `jam create-stack` requires the two `Dockerfile`s on disk to build the images. We could explore a way that it pulls an existing image rather than rebuilding a new one, but that's a feature that doesn't currently exist.

Beyond that, we could use symlinks (e.g. `../../../../../jammy-tiny-stack/stack/build.Dockerfile` but I think that's a really bad solution.

Reply:
Yea, the build image Dockerfile could possibly have `FROM paketobuildpacks/build-jammy-tiny` maybe in its Dockerfile.

Reply:
Yes, if `jam create-stack` were able to correctly update the metadata. Which I'm not sure it can do in the general case. What is a reasonable scope of existing metadata for `jam` to know about and overwrite? E.g. if the previous cnb user had UID of 1002 at build time and the new image should be 1004, how would `jam` know to remove the old one?

The more I think about it, the less in favor of this feature I am. I think it's reasonable to keep the `stack.toml` and `Dockerfile`s in sync some other way and not modify `jam` at all.

Reply:
I agree with @robdimsdale here. However, I tried to [start a conversation](https://cloud-native.slack.com/archives/C033DV9EBDF/p1662042914368639) with upstream CNB if we could loosen the cardinality of stacks (in particular run images) and builders somehow.

I can imagine there are huge overlaps between tiny and base build image as well. Wouldn't it be great to have an Ubuntu Jammy Stack with multiple build and run images and builders could select one of the build images, but different run images per e.g. order group.

We could reduce the number of stacks we need to maintain drastically, i.e. to a single one and potentially even the number of builders (with the use of extensions we might even be able to bring it down to a single builder). It would need a change of the CNB spec though. WDYT? I'd be very interested if this is just impossible to do / has been thought but was dropped / crazy thoughts or if it would be worth pursuing?

Reply:
I like this idea of optimizing for a shared `build` image with tailored `run` images though I haven't fully thought through the ramifications yet.

We could do this today at the builder level by having multiple builders, each with different `run` images that share the same stack ID, but I guess your point is that having something more clear in the spec could make this more discoverable and extensible.

Reply:
Whether we pursue that idea or not, I don't think we should block this RFC on that conversation